setup.py

```diff
@@ -74,7 +74,7 @@ def get_long_description() -> str:
     install_requires=[
         "click>=7.1.2",
         "appdirs",
-        "toml>=0.10.1",
+        "tomli>=0.2.6",
```

Comment:
```suggestion
        "tomli>=0.2.6,<2.0.0",
```
I would much appreciate this upper bound for the version. There's many open source projects afraid to make any changes, ever, due to the fear of downstream projects breaking. Luckily we have a solution: semantic versions. Let's use them :wink:

Reply:
I get the goals here, but I feel this kind of pinning is counterintuitive and causes unneeded stagnation. A recent example here is "myst-parser" blocking a sphinx upgrade on another project I help maintain: https://github.com/pypa/bandersnatch/pull/919
- I went and looked why and it's some formatting / mathematical syntax that's blocking it - i.e. it would work fine for 99.9999% of people ...

I feel packages should give people the two options version freedom and also offer pinned version. Thus I feel black should not do the '<2.0.0' here. Users of libraries should run their own CI to prove the dependencies works for them. Nice packages should also offer pinned (package==X/X) requirement files ([requirements.txt](https://github.com/pypa/bandersnatch/blob/master/requirements.txt)) or container images with tested versions for people who want to go that route. With Black, we do this via docker releases that give you the whole dependency tree pinned.

Reply:
> Users of libraries should run their own CI to prove the dependencies works

I agree 100%. Generally the rule should be that libraries allow unbounded dep versions (unless something is known not to work) and end-user applications should pin. My understanding is that Black is an end-user application, not a library (although can be used as one with no guarantees). Most users are installing via `pip install`, not using Docker, and will probably not be very happy when an upstream project makes a perfectly valid semantic version major version bump, and suddenly the `pip install black` is broken.

As to why myst-parser does the pinning, the reason is exactly the same: it considers itself an application first and foremost, not a library.

Also note that `tomli`'s public API is literally two functions and one error. If anything there changes, it is a thousand times more likely for something to break downstream than in the myst-parser example.

If you still disagree, I can totally respect that, but like said would appreciate the upper bound. I find it unlikely I'll ever tag a tomli v2.0.0 but if that does happen, these pins are incredibly nice to have.

Reply:
My opinion here is that while I also disagree semantic versioning does a good job at avoiding unnecessary breakage (https://hynek.me/articles/semver-will-not-save-you/) I don't think adding an upper bound for tomli is out of the question. While it would be the first upper-bound Black specified, I think giving tomli some room to change is fine (we hope that it doesn't but who knows what will happen) given the points hukkin has layed out. Another point is that since tomli is so focused, it's not *that* bad to be unable to upgrade. The only major exception is that brand new TOML syntax will be blocked once their support has been implemented in 2+, but assuming a version major bump does mean a broken API (of which Black uses 2/3 of) that's kinda moot :)

Reply:
> https://hynek.me/articles/semver-will-not-save-you/

The individual teachings in this write up make sense, but the main point seems to be "if you do things poorly, then the result is bad", :smile: from which I'm not sure how much there is to learn.

**cooperlees:**
> (pinning) causes unneeded stagnation

The way I see it, some stagnation happens regardless (unless nobody cares or pins in which case there is no stagnation but a lot of breakage). If application maintainers don't make the effort of testing and bounding to major versions then libraries stagnate in the fear of breaking user facing apps. I think the game is more fair if everyone takes care of their own stagnation and doesn't inflict it on others.

**ichard26:**
> I think giving tomli some room to change is fine

Related to what I wrote above, I don't think applications should have the power to _give_ libraries freedom to change. That freedom (with semantic version constraints) should be the default.

**TL;DR**
Just me expressing my views. I don't think this is that important, but arguing is fun :smile:

Reply:
> Related to what I wrote above, I don't think applications should have the power to give libraries freedom to change. That freedom (with semantic version constraints) should be the default.

Well okay, clearly that phrase of mine ended up being more of a hot take than I intended :slightly_smiling_face: The main point is that I don't think semantic versioning works as well as claimed. I do agree libraries should be free to change regardless of what versioning scheme they choose, and while I still don't really think semantic versioning is ideal, it is what tomli chose, and this is how their freedom is maintained so OK, you've convinced a little bit :smiley: